docs/website/contents/for-developers/HardWonWisdom.md

```diff
@@ -348,3 +348,165 @@ The key corresponding parts of the implementation are as follows.
+  The core idea is that an `[F, G]` chain should still specify a finite safe zone for `G` despite it currently being the last era in the sequence, since the blocks created by nodes running `[F, G]` will eventually be validated by future nodes running `[F, G, H ...]`.
+- Use IMMEDIATE only for eras that ended in the past or could end on the chain today (ie next era's implementation is already deployed).
+  For example this could be `F` in `[F, G]` but not `G`, since there's not yet a next era in the implementation.
+- Use NEVER only for what is known to be the final era the chain will ever have.
+  For example this could be `G` in `[F, G]` but not `F`, but only if the plan is for this chain to never transition out of `G`.
+  It's always safe to use the EXTENSIBLE trajectory instead of the NEVER trajectory, but the NEVER trajectory will allow the HFC to predict eg slot-and-time translations into the indefinite future.
+
+The HFC interface currently requires the user to specify the `EraParams`'s safe zone and the `singleEraTransition` function separately.
+As such, it is possible to accidentally specify an indefinite safe zone alongside a `singleEraTransition` that sometimes returns `Just`.
+It's also possible to accidentally use the NEVER trajectory for what is not actually the final era.
+These violations do not necessarily lead to irrevocable disaster, but they are likely to cause confusion and/or require eg some patches to the the ledger rules of preceding eras when releasing a new one (eg see PR <https://github.com/input-output-hk/cardano-ledger/pull/2785>).
+
+### Eras that contain no slots
+
+The HFC interface and contract together permit `singleEraTransition cfg eraParams start st` to return `Just (boundEpoch start)`.
+In that case, the era contains no slots on the chain.
+Since it therefore contains no blocks, `st` must be the era's initial ledger state, and so this equality rules out the EXTENSIBLE trajectory.
+It also cannot be the NEVER trajectory, since `singleEraTransition` is returning a `Just` at all.
+Therefore this equality is only possible for an era using the IMMEDIATE trajectory.
+
+As explained above, the safe zone is trivially satisfied for the IMMEDIATE-INDEPENDENT and IMMEDIATE-DEPENDENT-BOUND cases.
+In the IMMEDIATE-DEPENDENT-TRANSLATION case, however, `singleEraTransition` cannot be applied without the initial ledger state of the era, and so the end can't be determined before that initial ledger state is.
+Since in general some chains in the preceding non-empty era will determine the start of this era before they determine its initial ledger state, the IMMEDIATE-DEPENDENT-TRANSLATION case only necessarily satisfies the safe zone when some prefix of the era sequence contains no slots.
+That's when the era's initial ledger state necessarily available as soon as the its start is determined (since it's just a sequence of translations starting from the chain's genesis "block").
```

Comment:
```suggestion
That's when the era's initial ledger state is necessarily available as soon as its start is determined (since it's just a sequence of translations starting from the chain's genesis "block").
```
(maybe also 'starting from the initial era's ledger state after the chain's genesis "block"')

Reply:
Done.